### ISSUE TITLE
Two incorrect inconsistent-return-statements violations

### ISSUE THREAD

**Issue by nedbat** [PyCQA/pylint#1794]
I tried to make a simpler test case, and could not.

### Steps to reproduce
1. In a Python 2.7 virtualenv
2. git clone https://github.com/nedbat/coveragepy.git
3. cd coveragepy/
4. git checkout d8288282fcd54ec849b3345e564b8c3260d10bb5
5. pip install -r requirements/dev.pip
6. make lint

Transcript of complete reproduction session: https://gist.github.com/nedbat/24a66a2a7777d3a62cdbcd615508624c

### Current behavior

Two functions are incorrectly flagged with inconsistent-return-statements:

    ************* Module coverage.config
    coverage/config.py:56: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
    ************* Module coverage.misc
    coverage/misc.py:141: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)

### pylint --version output

    Using config file /private/tmp/coveragepy/pylintrc
    pylint 1.8.1,
    astroid 1.6.0
    Python 2.7.10 (default, Feb  7 2017, 00:08:15)
    [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]

These are the functions that are flagged:

    def expensive(fn):
        """A decorator to indicate that a method shouldn't be called more than once.

        Normally, this does nothing.  During testing, this raises an exception if
        called more than once.

        """
        if env.TESTING:
            attr = "_once_" + fn.__name__

            def _wrapped(self):
                """Inner function that checks the cache."""
                if hasattr(self, attr):
                    raise AssertionError("Shouldn't have called %s more than once" % fn.__name__)
                setattr(self, attr, True)
                return fn(self)
            return _wrapped
        else:
            return fn                   # pragma: not test

and:

    def options(self, section):
        for section_prefix in self.section_prefixes:
            real_section = section_prefix + section
            if configparser.RawConfigParser.has_section(self, real_section):
                return configparser.RawConfigParser.options(self, real_section)
        raise configparser.NoSectionError

**Comment by PCManticore** [comment]
Thanks! I can reproduce them as well.

**Comment by ddaanet** [comment]
First function can be reduced to:

```
def function(flag):
    # pylint: disable = no-else-return
    if flag:
        def _inner():
            return
        return None
    else:
        return None
```

The inner function must have a return statement to reproduce inconsistent-return-statements.

Fixing no-else-return also fixes inconsistent-return-statements.

The second function can be reduced to:
```
import socket

def options(flag):
    if flag:
        return None
    raise socket.error
```

The exception must not be a built-in exception to reproduce inconsistent-return-statement. It does not work with `os.error`.

**Comment by mthuurne** [comment]
What is interesting is that if in the first simplified test case the inner function is changed to return an expression, the `inconsistent-return-statements` message does not go away:
```python
def function(flag):
    # pylint: disable = no-else-return
    if flag:
        def _inner():
            return None
        return None
    else:
        return None
```
So it seems it is not just a mix-up of the return statements from the inner and outer function.

**Comment by hippo91** [comment]
@ddaanet your second function with socket.error doesn't raise any  ``inconsistent-return-statements`` message in my environment. 
Even the function you described in #1782 doesn't emit any message.
For example: 

```
$ cat bug_1782_bis.py 
#pylint: disable=missing-docstring
from socket import error

def options(flag):
    if flag:
        return None
    raise error

def return_error():
    return error

def raise_error(flag):
    if flag:
        return None
    raise return_error()
```

```
$ python3.5 ../pylint/bin/pylint bug_1782_bis.py 
No config file found, using default configuration

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

Can you check with a recent checkout of pylint please?

**Comment by nedbat** [comment]
@hippo91 thanks for looking into this.  The latest master of pylint still has this problem.  Here are simplified instructions for reproducing it with master of pylint (use a Python 2.7 virtualenv):

```
git clone https://github.com/nedbat/coveragepy.git
cd coveragepy/
git checkout d8288282fcd54ec849b3345e564b8c3260d10bb5
sed -i .bak 's;pylint==1.8.1;git+https://github.com/PyCQA/pylint.git@master#egg=pylint==0.0;' requirements/dev.pip
pip install -r requirements/dev.pip
python -m pylint coverage/config.py coverage/misc.py
```

**Comment by hippo91** [comment]
@nedbat thanks for this quick answer. For the ``expensive`` method, it should be ok after PR #1846 has been merged.
For the other false positive, the problem seems to come from astroid.
With python3.5 you should not have this problem.
When i lint the small example i exposed before:
  - with python3.5 : no message emitted
  - with python2.7 : 2 ``inconsisted-return-statements`` messages are emitted.

With python2.7, astroid is unable to infer the exception raised.
I'll have a look as this ASAP.
Thanks again.

**Comment by nedbat** [comment]
@hippo91 Ah, I overlooked that your PR hadn't been merged yet.  I tried using a Python 3.6 virtualenv, and used this changed sed line to get your PR:
```
sed -i .bak 's;pylint==1.8.1;git+https://github.com/hippo91/pylint.git@4771cbaf8f13fa79bd5ceda93990771297ac19d8#egg=pylint==0.0;' requirements/dev.pip
```
Now the pylint output is:
```
************* Module coverage.config
coverage/config.py:33: Parameters differ from overridden 'read' method (arguments-differ)
coverage/config.py:56: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
```
So one of the mistakes is fixed, and the other is not.

**Comment by nedbat** [comment]
Hmm, not sure what happened: I tried this with the just-released pylint 1.8.2, and I still have both incorrect inconsistent-return-statements violations:
```
lint recreate: /Users/ned/coverage/trunk/.tox/lint
lint installdeps: -rrequirements/dev.pip
lint develop-inst: /Users/ned/coverage/trunk
lint installed: apipkg==1.4,astroid==1.6.1,backports.functools-lru-cache==1.4,certifi==2018.1.18,chardet==3.0.4,configparser==3.5.0,-e hg+ssh://hg@bitbucket.org/ned/coveragepy@20ba4848d893ebbab1561b95e93b3dae8cb76e15#egg=coverage,decorator==4.2.1,enum34==1.1.6,execnet==1.5.0,flaky==3.4.0,funcsigs==1.0.2,greenlet==0.4.12,idna==2.6,isort==4.2.15,lazy-object-proxy==1.3.1,mccabe==0.6.1,mock==2.0.0,pbr==3.1.1,pkginfo==1.4.1,pluggy==0.6.0,py==1.5.2,PyContracts==1.8.0,pyenchant==2.0.0,pylint==1.8.2,pyparsing==2.2.0,pytest==3.2.5,pytest-forked==0.2,pytest-xdist==1.20.1,requests==2.18.4,requests-toolbelt==0.8.0,singledispatch==3.4.0.3,six==1.11.0,tox==2.9.1,tox-battery==0.5,tqdm==4.19.5,twine==1.9.1,unittest-mixins==1.4,urllib3==1.22,virtualenv==15.1.0,wrapt==1.10.11
lint runtests: PYTHONHASHSEED='599418143'
lint runtests: commands[0] | python -m pylint --notes= coverage tests igor.py setup.py __main__.py
Using config file /Users/ned/coverage/trunk/pylintrc
************* Module coverage.config
coverage/config.py:56: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
************* Module coverage.misc
coverage/misc.py:141: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
```

**Comment by nedbat** [comment]
Oh, the fix from @hippo91 is on master, but not on 1.8?  :(

**Comment by PCManticore** [comment]
Hey @nedbat sorry about that! It must have slipped the cherry-picking process.
@hippo91 when you get a chance can you push the said fix in 1.8 as well? we can release 1.8.3 with the new changes

**Comment by hippo91** [comment]
Ooooups sorry for that! I will backport it on 1.8 ASAP.

**Comment by hippo91** [comment]
@nedbat I just pushed the fix for the false positive emission of inconsistent-return-statements message when an inner function is defined under an if statement, in the 1.8 branch. 
Tell me if you see some improvements.

**Comment by nedbat** [comment]
@hippo91 Thanks, that does fix one of my two incorrect violations.  The other persists:
```
************* Module coverage.config
coverage/config.py:56: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
```

**Comment by hippo91** [comment]
Thanks @nedbat! I will investigate this ASAP.

Le 25 janv. 2018 8:51 PM, "Ned Batchelder" <notifications@github.com> a
écrit :

**Comment by hippo91** [comment]
@nedbat I found a simple function that reproduces your pylint error:

```
try:
    import ConfigParser as configparser
except ImportError:
    import configparser

def bug_1794(a):
    for x in range(a):
        if x == 100:
            return a
    raise configparser.NoSectionError('toto')
```

The false positive comes from the fact that ``astroid`` is not able to safely infer the raised exception.
This is due to the way the ``configparser`` module is imported. When ``astroid`` try to infer the attribute ``NoSectionError`` of the ``configparser`` module, it's as if it was trying to infer for both modules (one for python2 (line no 3) and one for python3 (line no 5)). 
Honestly i don't know how i can safely modify ``astroid`` to make it manage this situation. However when i import ``configparser`` with ``six`` module,  the problem disappears.

```
from six.moves import configparser

def bug_1794(a):
    for x in range(a):
        if x == 100:
            return a
    raise configparser.NoSectionError('toto')
```

Is it possible for you to import this module in ``backward.py`` with ``six``?

**Comment by nedbat** [comment]
I appreciate the difficulty of inferring details of Python code. It continues to astound me what pylint *can* infer.

Why does pylint need to infer anything about the value in the raise statement?  The fact that it's a raise statement means that whatever follows must of course be an exception, right?

**Comment by hippo91** [comment]
Yes of course you are right, but in order to know if the exception is caught by an handler, we need to know the type of exception. That's why we ask ``astroid`` to infer the type of exception.
For example in 

```
def returns_in_exceptions():
    try:
        raise ValueError('test')
    except ValueError:
        return 1
    except (OSError, TypeError):
        return 2
```
there is no reason to emit an inconsistent return because the exception is handled by the first ``except`` statement (and it returns explicitly).

**Comment by nedbat** [comment]
True, in that case, it needs to understand the type of the exception, as it does for other reasons.  But for my function, there's no reason to know anything more than, "this is raising an exception, and no statement after it will be executed, including the implicit return-None."

**Comment by hippo91** [comment]
@nedbat you are absolutely right. I will add a test to trigger the exception inference only if it is in a try/except statement. Otherwise there are no reason, i think, to infer it.
Thanks a lot!

**Comment by hippo91** [comment]
@nedbat it should be ok now if you update from 1.8 branch!
Let me know if the problem has disappeared!

**Comment by nedbat** [comment]
Indeed, using commit 91f8f1bea24b9a6933f281ad45b42fd2c8f89fef, the problems are gone.  Thanks! :)

**Comment by hippo91** [comment]
You are welcome @nedbat !
I'm closing this issue.
